File: python/tests/test_ocean_refraction.py
"""Tests for OceanRefractionProfile integration class."""

import math
import numpy as np
import pytest

from refraction.ocean.ocean_refraction import (
    OceanConditions, OceanMode, OceanRefractionProfile,
    OceanRefractivityResult, VectorOceanRefractivityResult,
)
from refraction.ocean.ocean_profile import FLORIDA_GULF_SUMMER, CALIFORNIA_SUMMER
from refraction.ocean import meissner_wentz as mw
from refraction.ocean import millard_seaver as ms
from refraction.ocean import quan_fry as qf


@pytest.fixture
def standard_cond(ocean_cond_radio):
    return ocean_cond_radio


@pytest.fixture
def standard_profile(standard_cond):
    return OceanRefractionProfile(standard_cond)


class TestCompute:
    """Single-depth computation."""

    def test_returns_result(self, standard_profile):
        r = standard_profile.compute(0.0, OceanMode.MeissnerWentz)
        assert isinstance(r, OceanRefractivityResult)

    def test_radio_surface_matches_standalone(self, standard_cond, standard_profile):
        r = standard_profile.compute(0.0, OceanMode.MeissnerWentz)
        n_standalone = mw.refractive_index(standard_cond.freq_ghz,
                                           standard_cond.sst_C,
                                           standard_cond.sss_psu)
        assert r.n_real == pytest.approx(n_standalone.n_real, rel=0.01)

    def test_optical_surface_matches_standalone(self, standard_cond, standard_profile):
        r = standard_profile.compute(0.0, OceanMode.QuanFry)
        n_standalone = qf.refractive_index(standard_cond.sst_C,
                                           standard_cond.sss_psu,
                                           standard_cond.wavelength_nm)
        assert r.n_real == pytest.approx(n_standalone, rel=0.01)

    def test_millard_seaver_surface(self, standard_cond, standard_profile):
        r = standard_profile.compute(0.0, OceanMode.MillardSeaver)
        n_standalone = ms.refractive_index(standard_cond.sst_C,
                                           standard_cond.sss_psu,
                                           0.0, standard_cond.wavelength_nm)
        assert r.n_real == pytest.approx(n_standalone, rel=0.01)

    def test_radio_has_permittivity(self, standard_profile):
        r = standard_profile.compute(50.0, OceanMode.MeissnerWentz)
        assert r.eps_real > 0.0
        assert r.eps_imag > 0.0

    def test_optical_no_imaginary(self, standard_profile):
        r = standard_profile.compute(50.0, OceanMode.QuanFry)
        assert r.n_imag == 0.0

    def test_dn_dz_computed(self, standard_profile):
        r = standard_profile.compute(50.0, OceanMode.MeissnerWentz)
        assert math.isfinite(r.dn_dz)


class TestProfile:
    """Depth profile computation."""

    def test_profile_length(self, standard_profile):
        results = standard_profile.profile(0.0, 100.0, 10.0, OceanMode.MeissnerWentz)
        assert len(results) == 11  # 0, 10, 20, ..., 100

    def test_profile_smooth(self, standard_profile):
        results = standard_profile.profile(0.0, 200.0, 5.0, OceanMode.MeissnerWentz)
        n_values = [r.n_real for r in results]
        # No NaN or Inf
        assert all(math.isfinite(n) for n in n_values)

    def test_millard_seaver_profile(self, standard_profile):
        results = standard_profile.profile(0.0, 200.0, 50.0, OceanMode.MillardSeaver)
        assert len(results) == 5
        # All values should be finite and physically reasonable (n > 1.3)
        for r in results:
            assert math.isfinite(r.n_real)
            assert r.n_real > 1.3


class TestCallable:
    """Scalar/array dispatch via __call__."""

    def test_scalar_call(self, standard_profile):
        r = standard_profile(50.0, OceanMode.MeissnerWentz)
        assert isinstance(r, OceanRefractivityResult)

    def test_array_call(self, standard_profile):
        r = standard_profile([0.0, 50.0, 100.0], OceanMode.MeissnerWentz)
        assert isinstance(r, VectorOceanRefractivityResult)
        assert len(r.depth_m) == 3


class TestIAPWSMode:
    """IAPWS mode through the profile class."""

    def test_iapws_surface(self, standard_profile):
        """IAPWS mode should return pure water RI (ignores salinity)."""
        r = standard_profile.compute(0.0, OceanMode.IAPWS)
        assert r.n_real > 1.3
        assert r.n_real < 1.4
        assert r.n_imag == 0.0

    def test_iapws_at_depth(self):
        """IAPWS with pressure: n should increase with depth."""
        cond = OceanConditions(sst_C=10.0, sss_psu=0.0, wavelength_nm=589.3)
        prof = OceanRefractionProfile(cond)
        r0 = prof.compute(0.0, OceanMode.IAPWS)
        r1000 = prof.compute(1000.0, OceanMode.IAPWS)
        assert r1000.n_real > r0.n_real


class TestPreset:
    """Preset ocean profiles."""

    def test_florida_preset(self):
        prof = OceanRefractionProfile.from_preset(FLORIDA_GULF_SUMMER)
        r = prof.compute(0.0, OceanMode.MeissnerWentz)
        assert r.temperature_C > 22.0  # tanh model doesn't reach exact SST

    def test_california_preset(self):
        prof = OceanRefractionProfile.from_preset(CALIFORNIA_SUMMER)
        r = prof.compute(0.0, OceanMode.MeissnerWentz)
        assert r.temperature_C > 14.0


class TestUserArrays:
    """User-supplied T(z), S(z) profiles."""

    def test_user_arrays(self):
        depths = [0.0, 50.0, 100.0, 200.0, 500.0]
        temps = [25.0, 20.0, 15.0, 8.0, 4.0]
        sals = [35.0, 35.0, 35.0, 34.5, 34.5]
        prof = OceanRefractionProfile.from_arrays(depths, temps, sals)
        r = prof.compute(0.0, OceanMode.MeissnerWentz)
        assert r.temperature_C == pytest.approx(25.0, abs=0.1)

    def test_interpolated_depth(self):
        depths = [0.0, 100.0]
        temps = [25.0, 5.0]
        sals = [35.0, 35.0]
        prof = OceanRefractionProfile.from_arrays(depths, temps, sals)
        r = prof.compute(50.0, OceanMode.MeissnerWentz)
        assert r.temperature_C == pytest.approx(15.0, abs=0.1)

File: cpp/tests/test_ocean_refraction.cpp
#include "test_utils.hpp"
#include <refraction/ocean/ocean_refraction.hpp>

using namespace refraction::test;
using namespace refraction::ocean;

static TestRunner runner("Ocean Refraction Integration Tests");

// ── Quan & Fry spot checks ──

TEST_CASE(runner, quan_fry_pure_water) {
    TEST_ASSERT_NEAR(qf::refractive_index(20.0, 0.0, 589.0), 1.333, 0.001);
}

TEST_CASE(runner, quan_fry_seawater) {
    TEST_ASSERT_NEAR(qf::refractive_index(20.0, 35.0, 589.0), 1.339, 0.001);
}

TEST_CASE(runner, quan_fry_dispersion) {
    TEST_ASSERT(qf::refractive_index(20, 35, 450) > qf::refractive_index(20, 35, 650));
}

// ── IAPWS spot checks ──

TEST_CASE(runner, iapws_check_value) {
    // IAPWS Table 3: T=25C, rho=997.047, lam=589.3nm -> n=1.33285
    double n = iapws::refractive_index(25.0, 589.3, 997.047);
    TEST_ASSERT_NEAR(n, 1.33285, 0.00002);
}

TEST_CASE(runner, iapws_density_4C) {
    TEST_ASSERT_NEAR(iapws::pure_water_density(4.0), 1000.0, 0.1);
}

TEST_CASE(runner, iapws_density_20C) {
    TEST_ASSERT_NEAR(iapws::pure_water_density(20.0), 998.2, 0.5);
}

// ── OceanRefractionProfile: radio mode ──

TEST_CASE(runner, radio_surface) {
    OceanConditions cond;
    cond.sst_C = 25.0;
    cond.sss_psu = 35.0;
    cond.freq_ghz = 10.0;
    OceanRefractionProfile prof(cond);
    auto r = prof.compute(0.0, OceanMode::MeissnerWentz);

    // Cross-check against standalone
    auto n = mw::refractive_index(10.0, 25.0, 35.0);
    TEST_ASSERT_NEAR_REL(r.n_real, n.n_real, 0.01);
    TEST_ASSERT(r.eps_real > 0.0);
    TEST_ASSERT(r.eps_imag > 0.0);
}

TEST_CASE(runner, radio_dn_dz_finite) {
    OceanConditions cond;
    cond.sst_C = 25.0;
    cond.sss_psu = 35.0;
    OceanRefractionProfile prof(cond);
    auto r = prof.compute(50.0, OceanMode::MeissnerWentz);
    TEST_ASSERT(std::isfinite(r.dn_dz));
}

// ── OceanRefractionProfile: optical modes ──

TEST_CASE(runner, millard_seaver_surface) {
    OceanConditions cond;
    cond.sst_C = 25.0;
    cond.sss_psu = 35.0;
    cond.wavelength_nm = 589.26;
    OceanRefractionProfile prof(cond);
    auto r = prof.compute(0.0, OceanMode::MillardSeaver);

    double n = ms::refractive_index(25.0, 35.0, 0.0, 589.26);
    TEST_ASSERT_NEAR_REL(r.n_real, n, 0.01);
    TEST_ASSERT_NEAR(r.n_imag, 0.0, 1e-15);
}

TEST_CASE(runner, quan_fry_surface) {
    OceanConditions cond;
    cond.sst_C = 20.0;
    cond.sss_psu = 35.0;
    cond.wavelength_nm = 589.0;
    OceanRefractionProfile prof(cond);
    auto r = prof.compute(0.0, OceanMode::QuanFry);

    double n = qf::refractive_index(20.0, 35.0, 589.0);
    TEST_ASSERT_NEAR_REL(r.n_real, n, 0.01);
}

// ── Profile sweep ──

TEST_CASE(runner, profile_length) {
    OceanConditions cond;
    cond.sst_C = 25.0;
    cond.sss_psu = 35.0;
    OceanRefractionProfile prof(cond);
    auto results = prof.profile(0.0, 100.0, 10.0, OceanMode::MeissnerWentz);
    TEST_ASSERT_EQ(static_cast<int>(results.size()), 11);  // 0,10,...,100
}

TEST_CASE(runner, profile_no_nan) {
    OceanConditions cond;
    cond.sst_C = 25.0;
    cond.sss_psu = 35.0;
    OceanRefractionProfile prof(cond);
    auto results = prof.profile(0.0, 200.0, 5.0, OceanMode::MeissnerWentz);
    for (const auto& r : results)
        TEST_ASSERT(std::isfinite(r.n_real));
}

// ── Preset profiles ──

TEST_CASE(runner, florida_preset) {
    OceanRefractionProfile prof(FLORIDA_GULF_SUMMER);
    auto r = prof.compute(0.0, OceanMode::MeissnerWentz);
    TEST_ASSERT(r.temperature_C > 22.0);
}

TEST_CASE(runner, california_preset) {
    OceanRefractionProfile prof(CALIFORNIA_SUMMER);
    auto r = prof.compute(0.0, OceanMode::MeissnerWentz);
    TEST_ASSERT(r.temperature_C > 14.0);
}

// ── User-supplied arrays ──

TEST_CASE(runner, user_arrays) {
    std::vector<double> depths = {0, 50, 100, 200, 500};
    std::vector<double> temps  = {25, 20, 15, 8, 4};
    std::vector<double> sals   = {35, 35, 35, 34.5, 34.5};
    OceanRefractionProfile prof(depths, temps, sals);
    auto r = prof.compute(0.0, OceanMode::MeissnerWentz);
    TEST_ASSERT_NEAR(r.temperature_C, 25.0, 0.1);
}

TEST_CASE(runner, user_arrays_interpolated) {
    std::vector<double> depths = {0, 100};
    std::vector<double> temps  = {25, 5};
    std::vector<double> sals   = {35, 35};
    OceanRefractionProfile prof(depths, temps, sals);
    auto r = prof.compute(50.0, OceanMode::MeissnerWentz);
    TEST_ASSERT_NEAR(r.temperature_C, 15.0, 0.1);
}

int main() {
    runner.run();
    return print_final_summary();
}
